tests/sync/app.cpp

```diff
@@ -1190,3 +1181,244 @@ TEST_CASE("app: response error handling", "[sync][app]") {
+
+        // Log in user 1
+        app.log_in_with_credentials(realm::app::AppCredentials::anonymous(),
+                                    [&](std::shared_ptr<realm::SyncUser> user, Optional<app::AppError> error) {
+            user_a = user;
+            CHECK(!error);
+        });
+        
+        CHECK(SyncManager::shared().get_current_user() == user_a);
+        
+        app.log_out([&](Optional<app::AppError> error) {
+            CHECK(!error);
+        });
+
+        CHECK(SyncManager::shared().get_current_user() == nullptr);
+        CHECK(user_a->state() == SyncUser::State::Removed);
+
+        // Log in user 2
+        app.log_in_with_credentials(realm::app::AppCredentials::anonymous(),
+                                    [&](std::shared_ptr<realm::SyncUser> user, Optional<app::AppError> error) {
+            user_b = user;
+            CHECK(!error);
+        });
+        
+        CHECK(SyncManager::shared().get_current_user() == user_b);
+        CHECK(SyncManager::shared().all_users().size() == 1);
+
+        app.switch_user(user_a, [&](std::shared_ptr<realm::SyncUser> user, Optional<app::AppError> error) {
+            CHECK(error->error_code.value() > 0);
+            CHECK(!user);
+        });
+        
+        CHECK(SyncManager::shared().get_current_user() == user_b);
+        
+        processed = true;
+        CHECK(processed);
+    }
+    
+}
+
+TEST_CASE("app: remove anonymous user", "[sync][app]") {
+    
+    std::function<std::unique_ptr<GenericNetworkTransport>()> transport_generator = [&] {
+        return std::unique_ptr<GenericNetworkTransport>(new UnitTestTransport());
+    };
+
+    auto config = App::Config{"translate-utwuv", transport_generator};
+    auto app = App(config);
+    
+    std::string base_path = tmp_dir() + "/" + config.app_id;
+    reset_test_directory(base_path);
+
+    auto tsm = TestSyncManager(base_path);
+    
+    bool processed = false;
+    std::shared_ptr<SyncUser> user_a;
+    std::shared_ptr<SyncUser> user_b;
+
+    SECTION("remove user expect success") {
+        CHECK(SyncManager::shared().all_users().size() == 0);
+
+        // Log in user 1
+        app.log_in_with_credentials(realm::app::AppCredentials::anonymous(),
+                                    [&](std::shared_ptr<realm::SyncUser> user, Optional<app::AppError> error) {
+            CHECK(!error);
+            CHECK(SyncManager::shared().get_current_user() == user);
+            user_a = user;
+        });
+        
+        CHECK(user_a->state() == SyncUser::State::LoggedIn);
+        
+        app.log_out(user_a, [&](Optional<app::AppError> error) {
+            CHECK(!error);
+            // a logged out anon user will be marked as Error, not LoggedOut
+            CHECK(user_a->state() == SyncUser::State::Removed);
+        });
+        
+        app.remove_user(user_a, [&](Optional<app::AppError> error) {
+            CHECK(error->message == "No user has been found");
```

Comment:
Is there a reason that removing an already removed user is an error? Couldn't it just be a no-op. I don't have a strong preference though.

But the error message feels a bit weird. Maybe rephrase to "User has already been removed"?

Reply:
I agree, a check has been added to see if they have already been removed